architecture/src/androidLibMain/kotlin/observable/Observable.kt

```diff
@@ -0,0 +1,135 @@
+
+ */
+@file:JvmName("AndroidObservable")
+
+package com.splendo.kaluga.architecture.observable
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MediatorLiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Observer
+import androidx.lifecycle.asLiveData
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.awaitCancellation
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import kotlin.coroutines.CoroutineContext
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.jvm.isAccessible
+
+actual interface WithState<T> {
+    actual val stateFlow:StateFlow<T>
+    actual val valueDelegate: ReadOnlyProperty<Any?, T>
+    val liveData: LiveData<T>
+        get()= stateFlow.asLiveData()
+}
+
+val <R:T, T, OO: ObservableOptional<R>> BasicSubject<R, T, OO>.liveDataObserver: Observer<T>
+    get() = Observer<T> { this.post(it) }
+
+private fun <B, R:T, T, OO: ObservableOptional<R>> B.mutableLiveData(): MutableLiveData<R> where B : BasicSubject<R, T, OO>, B : WithMutableState<R> {
+    val mediatorLiveData = MediatorLiveData<R>()
+    mediatorLiveData.addSource(stateFlow.asLiveData()) { value ->
+        mediatorLiveData.postValue(value)
+    }
+
+    return mediatorLiveData
+}
+
+fun <T> LiveData<T>.observeOnCoroutine(
+    coroutineScope: CoroutineScope,
+    coroutineContext: CoroutineContext = Dispatchers.Main.immediate,
+    observer: Observer<T>
+) {
+    coroutineScope.launch(coroutineContext) {
+        observeForever(observer)
+        awaitCancellation()
+    }.invokeOnCompletion {
+        removeObserver(observer)
+    }
+}
+
+actual abstract class BaseSubject<R:T, T, OO : ObservableOptional<R>> actual constructor(
+    observation: Observation<R, T, OO>,
+    stateFlowToBind:()-> StateFlow<R?>
+): AbstractBaseSubject<R, T, OO>(observation, stateFlowToBind) {
+
+    private lateinit var coroutineScope: CoroutineScope
+    val mutableLiveData: MutableLiveData<R> by lazy {
+        createLiveData().also {
+            if (this::coroutineScope.isInitialized) {
+                it.observeOnCoroutine(coroutineScope, observer = liveDataObserver())
+            }
+        }
+    }
+    protected abstract fun createLiveData(): MutableLiveData<R>
+    abstract fun liveDataObserver(): Observer<R>
+
+    final override fun bind(coroutineScope: CoroutineScope, context: CoroutineContext) {
+        super.bind(coroutineScope, context)
+        this.coroutineScope = coroutineScope
+        this::mutableLiveData.let { liveDataProperty ->
+            val accessibility = liveDataProperty.isAccessible
+            liveDataProperty.isAccessible = true
+            if ((liveDataProperty.getDelegate() as Lazy<*>).isInitialized()) {
+                mutableLiveData.observeOnCoroutine(coroutineScope, observer = liveDataObserver())
+            }
+            liveDataProperty.isAccessible = accessibility
```

Comment:
This complicated use of reflection / force casting is not really needed, if you store the lazy in a separate val:

 val:
```kotlin
val lazyDelegate = lazy { ... }
val value by lazyDelegate
lazyDelegate.isInitialized
```